features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+    }
+  }
+
+  if (done) {
+    return;
+  }
+
+  const authenticators = await skygear.auth.mfa.listAuthenticators();
+  // Present the authenticators to the user and let them choose
+  // which one they want to use.
+
+  // Suppose they chose OOB
+  await skygear.mfa.triggerOOB(authenticators[0].id);
+
+  // Present an UI to instruct the user to check SMS.
+  // Present an UI to offer the user to trigger resend.
+  // Present an UI to instruct the user to input the received code.
+  // Present an UI to offer the user to skip MFA for 30 days.
+
+  const code = textInput.value;
+  const { bearerToken } = await skygear.auth.mfa.authenticateWithOOB({ authenticatorID, code }, { requestBearerToken: true});
+
+  // Persist the bearer token
+  if (bearerToken) {
+    await storeBearerToken(bearerToken);
+  }
+}
+```
+
+The above example demonstrates the flow of authenitcation in details.
+The bearer token should be handled transparently by the SDK so the actual code written by the developer should not need to be aware of bearer token.
+
+## List authenticators in settings screen
+
+```typescript
+const authenticators = await skygear.auth.mfa.listAuthenticators();
+// [
+//   {
+//     id: "1",
+//     type: "totp",
+//     activatedAt: new Date("2019-07-19T00:00:00.000Z"),
+//     displayName: "totp-2edfdca2-3f05-41d2-bfd8-502725ee09a6",
+//   },
+//   {
+//     id: "2",
+//     type: "oob",
+//     activatedAt: new Date("2019-07-19T00:00:00.000Z"),
+//     channel: "sms",
+//     phone: "+85223456789",
+//   },
+//   {
+//     id: "3",
+//     type: "recovery_code",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//   },
+//   {
+//     id: "4",
+//     type: "recovery_code",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//   },
+//   {
+//     id: "5",
+//     type: "recovery_code",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//   },
+//   {
+//     id: "5",
+//     type: "recovery_code",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//   },
+//   {
+//     id: "6",
+//     type: "bearer_token",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//     expireAt: new Date("2019-08-19T00:00:00.000Z"),
+//   }
+// ]
+
+const numberOfRecoveryCodes = authenticators.filter(a => a.type === "recovery_code").length;
+const hasTOTP = authenticators.filter(a => a.type === "totp").length > 0;
+const hasOOB = authenticators.filter(a => a.type === "oob").length > 0;
+```
+
+## Delete an authenticator
+
+```typescript
+const authenticators = await skygear.auth.mfa.listAuthenticators();
+// Present an UI to show the list of authenticators and
+// let the user to choose which one to delete.
+
+// Suppose the user wants to delete the first one.
+const { id } = authenticators[0];
+await skygear.auth.mfa.deleteAuthenticator(id);
+
+// The user can also delete a bearer token if they believe
+// the device storing it has been compromised.
+const bearerTokens = authenticators.filter(a => a.type === "bearer_token");
```

Comment:
I don't think we need to provide a way to invalidate a specific bearer token. I just checked google only has option for revoking all trusted devices.

Reply:
So we should have a first-class function like `deleteAllBearerTokens()`? In the future when we design the session API, should we delete the associated bearer token when a session is revoked?

Reply:
Yes, I think a function named `revokeAllBearerTokens` is appropriate. If a session is revoked, I don't think we need to revoke the associated bearer token. However, if the API is 'logout this device', then we should revoke the associated bearer token.